types/events.d.ts

```diff
@@ -1,95 +1,104 @@
 export type EventType =
```

Comment:
Can't we use the native `HTMLElementEventMap` instead of maintaining a custom type?

Reply:
Didn't know this existed. I'd prefer this :+1:

Reply:
Yeah, `add.addEventListener` is typed, so my first thought is that we should be able to reuse those types.

```ts
// HTMLDocument
addEventListener<K extends keyof DocumentEventMap>(type: K, listener: (this: Document, ev: DocumentEventMap[K]) => any, options?: boolean | AddEventListenerOptions): void;
addEventListener(type: string, listener: EventListenerOrEventListenerObject, options?: boolean | AddEventListenerOptions): void;

// HTMLElement
addEventListener<K extends keyof HTMLElementEventMap>(type: K, listener: (this: HTMLElement, ev: HTMLElementEventMap[K]) => any, options?: boolean | AddEventListenerOptions): void;
addEventListener(type: string, listener: EventListenerOrEventListenerObject, options?: boolean | AddEventListenerOptions): void;
```

@ghostd, please correct me if I'm wrong. I might miss something fundamental.

Reply:
From my understanding, removing the custom `EventType` will not prevent us to maintain the event map (since the metadata are 'erased' at runtime), and for now, the lib.dom.ts is still missing some events ([composition ones](https://github.com/microsoft/TypeScript/issues/35634#issuecomment-636840255) and maybe others).
Maybe we can have a custom type which would be a union of the `DocumentEventMap`, `HTMLElementEventMap` and the missing events used by `dom-testing-library`. In this way we will just maintain the missing types. I'll give a shot

Reply:
I think i misunderstood your point.

We can simply define `EventType` this way:
```
type EventType = keyof (typeof eventMap);
```
When we add an entry into `eventMap`, TS will infer the new definition of `EventType`

Does this answer to your point?

Reply:
Ok, it seems i were wrong. The best i can get for now is:
```
type EventType = keyof HTMLElementEventMap | keyof WindowEventMap;
```
Then i can use this type to define the `eventMap`:
```
const eventMap: Record<EventType, EventMapValue> = {...all init values...}
```
There are 2 consequences:
* I have to change the key to be all lowercase (eg. `dblClick` -> `dblclick`). That does not seems to be a breaking change because `events.js` already mades a text transform:
https://github.com/testing-library/dom-testing-library/blob/b23a2bc8a2c570d8faaa76694e05bf781bbe36a0/src/events.js#L25
* the future `EventType` will be larger than the present one, it contains some key the library don't use, eg: `devicelight`,  `canplay`,  `volumechange`... I guess this is not a big deal

What do you think about this?

Reply:
I think I would prefer this solution. 

@eps1lon, what do you think?

Reply:
Sounds good to me. You two seem to know more about built-in types for this problem so I can't help much.

Reply:
I found this as an example from the documentation:
```
fireEvent.keyDown(domNode, { key: 'Enter', code: 'Enter' })
```
The case of the keys of `eventMap` seems to be part of the public API. Changing this would be a breaking change. I propose to have `keydown` in `eventMap` (so the key will be type checked against the lib.dom.ts interface), and add `keyDown` into `eventAliasMap`

I'll do a commit this day to show the changes.

Reply:
I've started something, and now i've to maintain the camelcase event list into the `eventAliasMap`:
```typescript
export const eventAliasMap: Record<string, EventType> = {
  doubleClick: 'dblclick',

  // Aliases to keep the API with camelcase events

  // Composition Events
  compositionEnd: 'compositionend',
  compositionStart: 'compositionstart',
  compositionUpdate: 'compositionupdate',
  // Keyboard Events
  keyDown: 'keydown',
  keyPress: 'keypress',
  keyUp: 'keyup',
  // Focus Events
  focusIn: 'focusin',
  focusOut: 'focusout',

  // Mouse Events
  contextMenu: 'contextmenu',
  dblClick: 'dblclick',
  dragEnd: 'dragend',
  dragEnter: 'dragenter',
  dragExit: 'dragexit',
  dragLeave: 'dragleave',
  dragOver: 'dragover',
  dragStart: 'dragstart',
  mouseDown: 'mousedown',
  mouseEnter: 'mouseenter',
  mouseLeave: 'mouseleave',
  mouseMove: 'mousemove',
  mouseOut: 'mouseout',
  mouseOver: 'mouseover',
// ...more and more...
}
```

We will to need an `AliasEventType` to correctly declare the all the fireEvent aliases (`fireEvent.dblClick`, `fireEvent.mouseEnter`, and so on). It seems we just shift the problem a little further.

The library API exposes camelcase events, which is more usable/readable for the end user IMHO. Maybe we should wait for the end of the TS migration before trying to simplify the code. For now, TS should be a "helper", not a constraint to build a clean API.

If you really want to see all the modifications i can spend more time on it this weekend. Let me know your opinions

Reply:
> If you really want to see all the modifications i can spend more time on it this weekend

Nah, let's not waste your time. I agree with:

> Maybe we should wait for the end of the TS migration before trying to simplify the code.

We've (you've) explored the available options. It looks like we have the best one for this moment.

So let get this thing merged 👍